add utility function to calculate underline size for category names

jacobgulan opened the pull request:
# Description

Fixes #626 

<!-- add a short summary if necessary; mention issue numbers -->
Added a helper function to calculate the underline size for the jinja templates. Specifically addresses not enough underline characters being produced for category names that include wide characters like emojis.


# Checklist
<!-- add a "X" inside the brackets to confirm -->
* [X] Make sure changes are covered by existing or new tests.
* [X] For at least one Python version, make sure test pass on your local environment.
* [X] Create a file in `src/towncrier/newsfragments/`. Briefly describe your
  changes, with information useful to end users. Your change will be included in the public release notes.
* [X] Make sure all GitHub Actions checks are green (they are automatically checking all of the above).
* [X] Ensure `docs/tutorial.rst` is still up-to-date.
* [X] If you add new **CLI arguments** (or change the meaning of existing ones), make sure `docs/cli.rst` reflects those changes.
* [X] If you add new **configuration options** (or change the meaning of existing ones), make sure `docs/configuration.rst` reflects those changes.

Commit:
add utility function to calculate underline size for category names

adiroiban commented on src/towncrier/_builder.py:
```diff
@@ -224,6 +241,11 @@ def split_fragments(
                 # it's recorded.
                 content = ""
 
+            # Calculate the underline size for the name of the category.
```
I don't think that this comment is needed.

The code is pretty clear.

```suggestion
```

jacobgulan replied:
Removed

adiroiban commented on src/towncrier/newsfragments/626.bugfix.rst:
```diff
@@ -0,0 +1 @@
+Fixed section names not having enough underline characters for emojis.
```
Maybe having this read like a release notes 

the current version looks like a commit message.

This information is for end users... and the end result is that they can now use emoji in category names :)

```suggestion
For reStructuredText format, you can now use emojis in category names.
In previous versions, the generated underline was not matching the title width. 
```

jacobgulan replied:
Done

adiroiban commented on src/towncrier/test/test_builder.py:
```diff
@@ -208,3 +208,13 @@ def test_ordering(self):
         )
+
+
+class TestUtilityFunctions(TestCase):
```
Very minor comment, but I think that we can have these tests in the same test case as the other tests.

------

These are unit tests, but we also need an integration/functional/end to end test to check that the default templates work.

jacobgulan replied:
Done

adiroiban commented on src/towncrier/_builder.py:
```diff
@@ -181,6 +182,22 @@ def find_fragments(
+    """
+    underline_size: int = 0
+    for char in text:
+        if unicodedata.east_asian_width(char) in ("W", "F"):
```
We also need a integration test.

If for example we have something like `if False:`
the existing tests will still pass with success. 

Please consider creating a test similar to this one, but in which a custom category is used, one that has unicode in the name.

In this way, we can make sure that this functionality will continue to work in future versions.

https://github.com/twisted/towncrier/blob/fc3a2556408bcdb0237e54c4ff0773db364c554a/src/towncrier/test/test_build.py#L1042-L1090

adiroiban reviewed (changes requested):
Change looks good.

To merge this we need mypy fixed and at least one integration/end to end test.

Many thanks for the PR

adiroiban commented on src/towncrier/_builder.py:
```diff
@@ -181,6 +182,22 @@ def find_fragments(
+    """
+    underline_size: int = 0
+    for char in text:
+        if unicodedata.east_asian_width(char) in ("W", "F"):
```
We also need a integration test.

If for example we have something like `if False:`
the existing tests will still pass with success. 

Please consider creating a test similar to this one, but in which a custom category is used, one that has unicode in the name.

In this way, we can make sure that this functionality will continue to work in future versions.

https://github.com/twisted/towncrier/blob/fc3a2556408bcdb0237e54c4ff0773db364c554a/src/towncrier/test/test_build.py#L1042-L1090

jacobgulan replied:
Added

adiroiban commented:
Note that the underline is also computed here ... line 237 


https://github.com/twisted/towncrier/blob/afa6cc0c5340df2fa2c10462f844a794e003cfdc/src/towncrier/build.py#L230-L239

Commit:
add integration test

Commit:
update underline size calc for jinja templates

Commit:
update bugfix newsfragment

Commit:
consolidate unit tests

jacobgulan commented:
Wanted to discuss the mypy change before touching it. It seems to be complaining that I'm trying to assign "underline_size" to a Mapping type. Mapping types are read-only. Therefore, mypy fails. I can update the type to a dict and mypy will likely be happy again, though I'm not sure that'd be the right approach.

adiroiban commented on src/towncrier/test/test_build.py:
```diff
@@ -1398,6 +1398,58 @@ def test_default_start_string_markdown(self, runner):
+    )
+    def test_underline_size_unicode(self, runner):
+        """
+        TODO: Add better description
```
Maybe something like this.
```suggestion
        The news file can be generaged for project names,
        and categories that contains emoji in their names.
        This is a test for generating the RST section
        underline to meet the docutils requirements. 
```

jacobgulan replied:
Whoops, totally forgot to come back around for that. Thoughts on my mypy comment above?

Commit:
update docstring

Commit:
update write_text() to use utf-8 encoding

adiroiban commented on src/towncrier/_builder.py:
```diff
@@ -246,6 +263,10 @@ def split_fragments(
                 # it's recorded.
                 content = ""
 
+            definitions[category]["underline_size"] = get_underline_size(
```
The big problem with this code, is that the underline size should not be computed inside the `split_fragments` function

also, maybe call it `underline_lenght` ? but this is a very minor comment

adiroiban reviewed (changes requested):
Thanks for the update. It's a good start.

Please consider using `render_fragments` to create the data for the jinja template.

Inside `render_fragments` we can create a copy of `definitions` and update that copy and then pass it to jinja template rendering

so we would have something like `definitions=updated_definitions` in the call below

https://github.com/twisted/towncrier/blob/4aa8174dc7e843e7b71fc3092216dc2915065135/src/towncrier/_builder.py#L405-L409


this should clean the code and should also make mypy happy

it's best not to randomly update the configuration object or to update the configuration object after it was loaded from the config file

Commit:
rename underline_size to underline_length

Commit:
pass get_underline_length() function to jinja template

Commit:
add more unit tests

Commit:
appease mypy generic types

Commit:
update typing for underline length

jacobgulan commented:
Appreciate the feedback. I switched directions on the implementation after your last comment and seeing a better way that should prove to be more reliable and maintainable.

Instead of creating a copy of definitions and then calculating the underline size for that modified dictionary, I'm instead just passing the jinja template the `get_underline_length()` function itself now and calculating the length with it inside the .rst files.

Added some helper functions to recursively handle other types like dicts & lists which proved necessary for some of the other integration tests, but the behavior is working as expected now without modifying the original definitions mapping.

Commit:
add TypeAlias

adiroiban commented:
Thanks for the update.

The data exposed to the jinja template is part of our public API.
We need to double check that it make sense, as deprecating this type of code is very hard.

--------

For a template design , I like the first way of exposing the underline lenght better.

As a general rule for template design, I try to avoid calling methods/functions from the template... but I can't tel if this is good or bad.
I feel that doing function calls in the templates, makes the template harder to read.

jacobgulan commented:
I personally think that reading a function is easier to follow - if the name is intuitive - than indexing off variables. I would say it also makes sense to consolidate calculating length to one function rather than mixing get_underline_length() & the jinja pipe operation.

That said, I can also see the argument to maintain what's already present and just add another field to it. That second approach would look something like this:
```
  updated_definitions = {
      category: {**value, "underline_length": get_underline_length(value["name"])}
      for category, value in definitions.items()
  }
```
Couldn't use copy.deepcopy() as mypy screams about passing the wrong argument.

Let me know what you think

adiroiban commented:
We can go with a separate function... and later we can look at making it available as a filter.

Right now, we use the jinja `Template()` API instead of the `Environment()` API and I don't know how to add an environment to an existing template... so that we can add a custom filter to the template.

I know that you can add custom filters to environments, but not to templates.

------

The thing is that jinja2 is a beast of its own... with inheritence, conditional statements, variable assignations ...  and I have seen a lot jinja2 templates that are very hard to understand and update.

This is why I prefer to go with the pre-computed view-model approach, so that in the template, I only have loops and simple variables insertion.

adiroiban commented on src/towncrier/templates/hr-between-versions.rst:
```diff
@@ -1,22 +1,22 @@
 {% if versiondata.name %}
 {{ versiondata.name }} {{ versiondata.version }} ({{ versiondata.date }})
-{{ top_underline * ((versiondata.name + versiondata.version + versiondata.date)|length + 4)}}
+{{ top_underline * (get_underline_length(versiondata.name + versiondata.version + versiondata.date) + 4)}}
```
One idea is a function like this

```diff
-{{ versiondata.name }} {{ versiondata.version }} ({{ versiondata.date }})
-{{ top_underline * (get_underline_length(versiondata.name + versiondata.version + versiondata.date) + 4)}}
+{{ top_section(versiondata.name, versiondata.version) }}
```

this function will join the arguments with a single space, and will generate 2 lines.

This looks to me less error prone

just an idea
